zerver/tests/test_auth_backends.py

```diff
@@ -152,6 +156,66 @@ def test_remote_user_backend_sso_append_domain(self):
             self.verify_backend(ZulipRemoteUserBackend(),
                                 email_to_username=email_to_username)
 
+    def test_github_backend(self):
+        email = 'hamlet@zulip.com'
+        good_kwargs = dict(response=dict(email=email), return_data=dict())
+        bad_kwargs = dict()  # type: Dict[str, str]
+        self.verify_backend(GitHubBackend(),
+                            good_kwargs=good_kwargs,
+                            bad_kwargs=bad_kwargs)
+
+class GitHubBackendTest(AuthedTestCase):
+    def setUp(self):
+        self.email = 'hamlet@zulip.com'
+        self.name = 'Hamlet'
+        self.backend = GitHubBackend()
+        self.backend.strategy = DjangoStrategy(storage=BaseDjangoStorage())
+        self.user_profile = get_user_profile_by_email(self.email)
+        self.user_profile.backend = self.backend
+
+    def test_github_backend_do_auth(self):
+        def do_auth(return_data=dict(), *args, **kwargs):
+            return self.user_profile
+
+        with mock.patch('zerver.views.login_or_register_remote_user') as result, \
+                 mock.patch('social.backends.github.GithubOAuth2.do_auth',
+                            side_effect=do_auth):
+            response=dict(email=self.email, name=self.name)
+            self.backend.do_auth(response=response)
+            result.assert_called_with(None, self.email, self.user_profile,
+                                      self.name)
+
+        def do_auth_inactive(return_data=dict(), *args, **kwargs):
+            return_data['inactive_user'] = True
+            return self.user_profile
+
+        with mock.patch('zerver.views.login_or_register_remote_user') as result, \
+             mock.patch('social.backends.github.GithubOAuth2.do_auth',
+                        side_effect=do_auth_inactive):
+            response=dict(email=self.email, name='Hamlet')
+            user = self.backend.do_auth(response=response)
+            result.assert_not_called()
+            self.assertIs(user, None)
+
+    def test_github_backend_new_user(self):
+        rf = RequestFactory()
+        request = rf.get('/complete')
+        request.session = {}
+        request.user = self.user_profile
+        self.backend.strategy.request = request
```

Comment:
So, full disclaimer, I've never written back end Django authentication code, so my inability to understand what's going on here might just reflect an unusual level of ignorance on my part, but...

I think we need a comment here to to explain what's going on.  Before we call `self.backend.do_auth()`, we need to attach `request` to `self.backend.strategy`.  Why doesn't `request` get passed in to some kind of action?